### DISCUSSION TITLE
Closing Connections

### DISCUSSION THREAD

**Discussion by icing** [curl/curl#11878] [Ideas]
I discovered several issues around closing connections after a transfer has been done. These popped up during handling of issues around TLS shutdown and test reliability in FTP.

The common theme here is that there is network IO during connection shutdown and we are not prepared to handle that properly. Also, these actions are completely ignored by an `-v` or `--trace`, so a user never sees any output from it.

### Current State

After transfers are done, connections land in the connection cache and get shut down on general cleanup. For running the cleanup, there is a generic, internal easy handle that gets attached to each connection to close. Since this is purely internal and does not inherit anything from the transport handles, no `DEBUG` callbacks are there. All `infof()` and `failf()` calls never produce any output.

The close of a connection is a one-shot call. If there is something to send, e.g. a TLS shutdown or a HTTP/2 GOAWAY, and the socket blocks, the data is never written. This most likely explains the spurious failures in FTP tests where the final `QUIT` command never reaches our test server.

The question is then: is this a conscious design decision or just by accident. Do we want to introduce a (short) shutdown period for connection close that will `poll()` blocked connections. Or not. Which is also a valid choice, but then we need to document this and adapt our test cases to it.

I believe we need at least make sure that the tool wil and applications can handle trace output.

**Comment by jay** [comment]
I commented in #11858 a few minutes ago , basically we know about this with TLS see #6149. We have these connections that `*may*` need to wait for a TLS shutdown phase, but we don't need to do that for almost all connections, and so right now we don't do it for any.

The FTPS case is really different because the upload needs to be completed and that separate connection closed before returning to the caller but here's a harder example imo to think about: curl_share_cleanup calls Curl_conncache_close_all_connections which closes all the connections, what happens if that blocks, should it block, because if we have to wait for all connections to shutdown then it's going to block. I don't have a good solution for this. HTTP/2 I don't remember us discussing the GOAWAY.

Here's another way to put it think about that above and your proposal in #11858. 1 second may not seem like a lot but you've got some time issues (ok could be solved by Curl_timeleft to handle user timeouts and Curl_timediff with timestamps to work around granularity), it's openssl specific and then aside from that you have to multiply by the amount of connections in the cache so question of scale.

**Comment by icing** [comment]
I will amend #11858 to not do that time wait, only shoot of the shutdown and if it fails to deliver, then that is so.

We have been living with that restriction for a while now. I do not think it was very wise of such protocols as TLS to demand a shutdown handshake. Serious servers cope with client just closing the socket anyway. But recent OpenSSL decided to change behaviour and gives now errors in such cases which server need to navigate around. Just FYI.

What I think we *should* address is that `curl -v` does not trace any messages during those shutdowns. That is information where the user, and ulitmately we in issue reports, fly blind now.

**Comment by dfandrich** [comment]
FWIW, [here's a log](https://ci.appveyor.com/project/curlorg/curl/builds/48218910/job/51e4orjrvxyvmv5q?fullLog=true) of a job of an IMAP test (test 1321) where the final `LOGOUT` command was not sent. The sockfilt log doesn't show any indication that curl actually sent the LOGOUT.

**Reply by mattjax16** [reply]
Very interesting

**Comment by icing** [comment]
Closing as outdated.
